# C10. Exception Handling

## 1. Exception-Handling Fundamentals

##### A Java exception is an object that describes an exceptional condition that has occurred in a piece of code.
when an exceptional condition arises, an object representing that exception is created and thrown in the method that caused the error.

##### managed via five keywords: try, catch, throw, throws, and finally.
- if an exception occurs within the **try** block, it is thrown. 
- **catch** the exception
- manully **throw** an exception, or Java run-time system automatically throw system-generated exceptions
- any exception that is thrown out of a method must be specified as such by a **throws** clause
- any code in a **finally** block executed after a try block

from JDK7, *try-with-resources*

## 2. Exception Types

##### All exception types are subclasses of the built-in class Throwable.
two subclasses:

1. Exception
    + *used for exceptional conditions that user programs should catch.*
    + *subclass to create custom exception types*
    + RuntimeException
        * subclass of Exception
        * automatically defined (e.g. division by zero)
2. Error
    + *are not expected to be caught under normal circumstances*
    + *Java run-time system indicate errors having to do with the run-time environment itself. (e.g. Stack overflow)*

## 3. Uncaught Exceptions

##### Any exception that is not caught by your program will ultimately be processed by the default handler.
The default handler displays a string describing the exception, prints a stack trace, and terminates the program.

## 4. Using try and catch

##### handling exception by user. benefits:
1. it allows you to fix the error
2. it prevents the program from automatically terminating.

##### try catch clause
> try {
> } catch (SomeException e) {
> }

### 4.1 Displaying a Description of an Exception

##### *Throwable* overrides the *toString()* method so that it returns a string containing a description of the exception.
> println(e);

## 5. Multiple catch Clauses

##### each *catch* statement is inspected in order. after one executes, the others are bypassed.
throw error when superclass before subclass during compiling

## 6. Nested try statements

##### the *try* statement can be nested.

## 7. throw

##### throw an exception explicitly
> throw ThrowableInstance;

*ThrowableInstance* must be an object of type **Throwable** or a subclass of **Throwable**.

1. using a parameter in a catch clause (e.g. throw e;)
2. creating one with the new operator (e.g. throw new Exception("error");)

##### The flow of execution stops immediately after the throw statement

## 8. throws

##### a throws clause is necessary for all exceptions which is not handled within the method, except those of type Error or RuntimeException, or any of their subclasses.
Otherwise, compile-time error

> type method-name(parameter-list) throws exception-list {

## 9. finally

##### *finally* creates a block of code that will be executed after a *try / catch* block has completed
- the finally block will execute whither or not an exception is thrown
- will execute even if no catch statement matches the exception
- finally clause is optional. however, each try statement requires at least one catch or a finally clause.

## 10. Java's Built-in Exceptions

Inside the standard package java.lang

In the language of Java
1. unchecked exceptions
    - the compiler does not check to see if a method handles or throws these exceptions (subclasses of the standard type RuntimeException)
2. checked exceptions
    - must be included in a method's throws list if that method does not handle it itself

## 11. Creating Your Own Exception Subclasses

##### *Exception* defines four public constructors. two support chained exceptions.
- Exception()
- Exception(String msg)

sometimes it is better to override toString() (defined by Throwable, displays the name of exception followed by a colon)

## 12. Chained Exceptions
Beginning with JDK 1.4

##### allows to associate another exception with an exception. the second describes the cause of the first exception
two constructors were added to Throwable (also Error, Exception, and RuntimeException)
- Throwable(Throwable causeExc)
- Throwable(String msg, Throwable causeExc)

chained exception methods
- Throwable getCause()
- Throwable initCause(Throwable causeExc)

```java
//e.g.
NullPointerException e = new NullPointerException("error");
e.initCause(new ArithmeticException("cause"));
```

## 13. Three Recently Added Exception Features
Beginning with JDK 7

1. try-with-resources
    - automates the process of releasing a resource when it is no longer needed.
2. multi-catch
    - a single *catch* clause to handle different exceptions - separate by |
    - Each multi-catch parameter is implicitly *final* - cannot be changed in catch
    - e.g. catch(ArithmeticException | ArrayIndexOutOfBoundsException e) {
3. final rethrow / more precise rethrow
    - restricts the type of exceptions that can be rethrown to only those checked exceptions that..
        + that the associated try block throws
        + that are not handled by a preceding catch clause
        + and that are a subtype or supertype of the parameter

## 14. Using Exceptions

Java's exception-handling statements should not be considered a general mechanism for nonlocal branching